providers/slurm/slurm.go

```diff
@@ -0,0 +1,532 @@
+func (p *Provider) UpdatePod(ctx context.Context, pod *v1.Pod) error {
+	log.Printf("Update pod %s", podName(pod.Namespace, pod.Name))
+	pi, ok := p.pods[podName(pod.Namespace, pod.Name)]
+	if !ok {
+		return ErrPodNotFound
+	}
+	pi.pod = pod
+
+	return nil
+}
+
+// DeletePod deletes pod.
+func (p *Provider) DeletePod(ctx context.Context, pod *v1.Pod) error {
+	log.Printf("Delete %s", podName(pod.Namespace, pod.Name))
+	delete(p.pods, podName(pod.Namespace, pod.Name))
+	return nil
+}
+
+// GetPod returns a pod.
+func (p *Provider) GetPod(ctx context.Context, namespace, name string) (*v1.Pod, error) {
+	log.Printf("Get Pod %s", podName(namespace, name))
+	pj, ok := p.pods[podName(namespace, name)]
+	if !ok {
+		return nil, ErrPodNotFound
+	}
+
+	return pj.pod, nil
+}
+
+// GetContainerLogs returns logs if requested pod is SlurmJob,
+// otherwise returns empty reader.
+func (p *Provider) GetContainerLogs(ctx context.Context, namespace, pName, containerName string, opts api.ContainerLogOpts) (io.ReadCloser, error) {
+	log.Printf("GetContainerLogs n:%s pod:%s containerName:%s", namespace, pName, containerName)
+
+	pi, ok := p.pods[podName(namespace, pName)]
+	if !ok {
+		return nil, ErrPodNotFound
+	}
+
+	if pi.jobID == 0 { // skipping non slurm jobs
+		return ioutil.NopCloser(strings.NewReader("")), nil
+	}
+
+	infoR, err := p.slurmAPI.JobInfo(ctx, &sAPI.JobInfoRequest{JobId: pi.jobID})
+	if err != nil && pi.jobInfo == nil {
+		return nil, errors.Wrap(err, "can't get slurm job info")
+	}
+
+	if infoR != nil {
+		pi.jobInfo = infoR.Info[0]
+	}
+
+	openResp, err := p.slurmAPI.OpenFile(context.Background(), &sAPI.OpenFileRequest{Path: pi.jobInfo.StdOut})
+	if err != nil {
+		return nil, errors.Wrap(err, "can't open slurm job log file")
+	}
+
+	buff := &bytes.Buffer{}
+	for {
+		c, err := openResp.Recv()
+		if c != nil {
+			_, err := buff.Write(c.Content)
+			if err != nil {
+				break
+			}
+		}
+
+		if err != nil {
+			break
+		}
+	}
+
+	return ioutil.NopCloser(buff), nil
+}
+
+// RunInContainer SLURM doesn't support it.
+func (p *Provider) RunInContainer(ctx context.Context, namespace, name, container string, cmd []string, attach api.AttachIO) error {
+	return ErrNotSupported
+}
+
+// GetPodStatus retrieves the status of a given pod by name.
+func (p *Provider) GetPodStatus(ctx context.Context, namespace, name string) (*v1.PodStatus, error) {
+	log.Printf("Get Pod Status namespace:%s name:%s", namespace, name)
+
+	status := &v1.PodStatus{
+		Phase:  v1.PodRunning,
+		HostIP: "1.2.3.4",
+		PodIP:  "5.6.7.8",
+		Conditions: []v1.PodCondition{
+			{
+				Type:   v1.PodInitialized,
+				Status: v1.ConditionTrue,
+			},
+			{
+				Type:   v1.PodReady,
+				Status: v1.ConditionTrue,
+			},
+			{
+				Type:   v1.PodScheduled,
+				Status: v1.ConditionTrue,
+			},
+		},
+	}
+
+	pod, err := p.GetPod(ctx, namespace, name)
+	if err != nil {
+		return nil, err
+	}
+
+	now := metav1.NewTime(time.Now())
+	if pod.Status.StartTime != nil {
```

Comment:
Isn't it always non-nil? Pod is added to map after `StartTime` is initialized

Reply:
Not always. In update pod can come with nil start time